### PR TITLE
Glacier Peak competition code

### PR DESCRIPTION
**Pull request by 1198159**

### CODE REVIEW

**Review by bobtseattle — COMMENTED** [review]
I am hesitant to push this through with so much disabled and commented out.  
We don't know which subsystems are really working and not given all the commented out code to get others tested.
Next time we should make sure to create a PR after each subsystem is completed.
The PR is up so Jonah is able to work off the branch.

File: src/main/java/frc/team2412/robot/Hardware.java [hunk 1]
```diff
@@ -142,28 +142,28 @@ public Hardware() {
         if (!comp)
             return;
         if (CLIMB_ENABLED) {
-            climbMotorDynamic = new WPI_TalonFX(CLIMB_DYNAMIC_MOTOR);
+            // climbMotorDynamic = new WPI_TalonFX(CLIMB_DYNAMIC_MOTOR);
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/Hardware.java:145]: This causes simulation to immediately crash. Need to actually update the climb subsystem to handle the optional dynamic arm vs. just commenting out.

File: src/main/java/frc/team2412/robot/Robot.java [hunk 2]
```diff
@@ -257,9 +259,18 @@ public void autonomousInit() {
     public void teleopInit() {
         if (SubsystemConstants.DRIVE_ENABLED) {
             subsystems.drivebaseSubsystem.setDefaultCommand(new DriveCommand(subsystems.drivebaseSubsystem,
-                    controls.driveController.getLeftXAxis(), controls.driveController.getLeftYAxis(),
+                    controls.driveController.getLeftYAxis(), controls.driveController.getLeftXAxis(),
                     controls.driveController.getRightXAxis()));
         }
+        if (!isCompetition())
+            return;
+        if (INTAKE_ENABLED) {
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/Robot.java:267]: Check whether intakeSubsystem is non-null vs. using another variable to infer if intakeSubsystem is valid or not.

File: src/main/java/frc/team2412/robot/Robot.java [hunk 3]
```diff
@@ -257,9 +259,18 @@ public void autonomousInit() {
...
+            subsystems.intakeSubsystem.intakeExtend();
+        }
+        if (SHOOTER_ENABLED) {
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/Robot.java:270]: Check if shooterSubsystem is non-null vs. using something else to infer if it's valid or not.

File: src/main/java/frc/team2412/robot/commands/shooter/ShooterTargetCommand.java [hunk 4]
```diff
@@ -6,25 +6,42 @@
...
+                .rotateBy(Rotation2.fromDegrees(shooter.getTurretAngle() + ShooterConstants.TURRET_ANGLE_OFFSET)).x
+                * ShooterConstants.TURRET_FF;
+        if (ShooterConstants.dataPoints != null) {
+            ShooterDataDistancePoint shooterData = ShooterConstants.dataPoints.getInterpolated(distance);
+            shooter.setHoodAngle(shooterData.getAngle());
+            shooter.setFlywheelRPM(shooterData.getRPM());
+
+        }
+
+        shooter.updateTurretAngle(yaw + ff);
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/commands/shooter/ShooterTargetCommand.java:44]: This fights with the IndexBitmapCommand that also sets the turret angle.

File: src/main/java/frc/team2412/robot/commands/shooter/ShooterTargetCommand.java [hunk 5]
```diff
@@ -6,25 +6,42 @@
...
     public void execute() {
         double distance = vision.getDistance() + shooter.getDistanceBias();
         double yaw = vision.getYaw() + shooter.getTurretAngleBias();
-        ShooterDataDistancePoint shooterData = ShooterConstants.dataPoints.getInterpolated(distance);
-        shooter.setHoodAngle(shooterData.getAngle());
-        shooter.setFlywheelRPM(shooterData.getRPM());
-        shooter.updateTurretAngle(yaw);
+        double ff = feedforward.get()
+                .rotateBy(Rotation2.fromDegrees(shooter.getTurretAngle() + ShooterConstants.TURRET_ANGLE_OFFSET)).x
+                * ShooterConstants.TURRET_FF;
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/commands/shooter/ShooterTargetCommand.java:36]: Move this after the `if()` statement block, as it's not used until the `updateTurretAngle()` call below.

File: src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java [hunk 6]
```diff
@@ -119,7 +119,7 @@ public ClimbSubsystem(WPI_TalonFX climbFixedMotor, WPI_TalonFX climbDynamicMotor
 
         climbFixedMotor.config_kP(PID_SLOT_0, P);
 
-        climbDynamicMotor.configAllSettings(motorConfig);
+        // climbDynamicMotor.configAllSettings(motorConfig);
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:122]: Instead, just check if it's non-null so you don't have to keep touching this code as the climb arm is added or removed.

Comment (jamesdooley4) [reply src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:122]: Ditto for elsewhere below & where the solenoid is used.

File: src/main/java/frc/team2412/robot/subsystem/DrivebaseSubsystem.java [hunk 7]
```diff
@@ -243,7 +242,8 @@ public double getAngularVelocity() {
 
     public Rotation2 getAngle() {
         synchronized (kinematicsLock) {
-            return Robot.getInstance().isCompetition() ? getPose().rotation.inverse() : getPose().rotation;
+            return getPose().rotation; // return Robot.getInstance().isCompetition() ? getPose().rotation.inverse() :
+                                        // getPose().rotation;
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystem/DrivebaseSubsystem.java:246]: Delete commented-out code, as I presume (and hope) it's never coming back.

File: src/main/java/frc/team2412/robot/subsystem/IndexSubsystem.java [hunk 8]
```diff
@@ -39,8 +30,8 @@
 
         // Index Motor Speeds
 
-        public static double INDEX_IN_SPEED = 0.1; // will change values later
-        public static double INDEX_OUT_SPEED = -0.1; // will also change later
+        public static double INDEX_IN_SPEED = 0.3; // will change values later
+        public static double INDEX_OUT_SPEED = -0.5; // will also change later
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystem/IndexSubsystem.java:34]: Can remove the comment, as the values were changed...

File: src/main/java/frc/team2412/robot/subsystem/IndexSubsystem.java [hunk 9]
```diff
@@ -237,70 +230,66 @@ public boolean feederHasCorrectCargo() {
...
+        // ingestOverCurrentStart = Timer.getFPGATimestamp();
+        // }
+        // }
+        // if (ingestCurrent > CURRENT_LIMIT_RESET_AMPS) {
+        // if (ingestOverCurrentStart > 0) {
+        // if (Timer.getFPGATimestamp() - ingestOverCurrentStart > CURRENT_LIMIT_TRIGGER_SECONDS) {
+        // ingestMotorStop();
+        // }
+        //
+        // }
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystem/IndexSubsystem.java:249]: Just delete vs. commenting out, as it looks like `SupplyCurrentLimitConfiguration` handles the same thing (though would continue to let balls be further jammed vs. stopping the motors entirely).

File: src/main/java/frc/team2412/robot/subsystem/IntakeSubsystem.java [hunk 10]
```diff
@@ -72,7 +73,7 @@ public IntakeSubsystem(WPI_TalonFX motor, DoubleSolenoid intakeSolenoid) {
         intakeSolenoidState = EXTEND;
         intakeMotorState = STOPPED;
 
-        intakeRetract();
+        // intakeRetract();
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystem/IntakeSubsystem.java:76]: Can leave uncommented, as `intakeRetract()` safely handles missing hardware (or should).

File: src/main/java/frc/team2412/robot/subsystem/IntakeSubsystem.java [hunk 11]
```diff
@@ -124,7 +125,7 @@ public void intakeStop() {
      */
     public void intakeExtend() {
         intakeSolenoidState = RETRACT;
-        solenoid.set(RETRACT.value);
+        // solenoid.set(RETRACT.value);
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystem/IntakeSubsystem.java:128]: Don't comment out - instead check if `solenoid` is non-null before using it, so you don't have to keep modifying this code.  Ditto for below.

File: src/main/java/frc/team2412/robot/subsystem/ShooterSubsystem.java [hunk 12]
```diff
@@ -84,8 +85,8 @@ public ShooterSubsystem(WPI_TalonFX flywheelMotor1, WPI_TalonFX flywheelMotor2,
 
         // Estimated gearing constant of 41
         public static final double TURRET_DEGREES_TO_ENCODER_TICKS = 41 * 2048 / 360; // 233
-        public static final double MIN_TURRET_ANGLE = -115; // Can barely reach -139 degrees physically 115 tested
-        public static final double MAX_TURRET_ANGLE = 245; // Can barely reach 210 degrees physically 245 tested
+        public static final double MIN_TURRET_ANGLE = -200; // Can barely reach -139 degrees physically 115 tested
+        public static final double MAX_TURRET_ANGLE = 115; // Can barely reach 210 degrees physically 245 tested
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystem/ShooterSubsystem.java:89]: Comments don't match the code anymore...

File: src/main/java/frc/team2412/robot/subsystem/ShooterSubsystem.java [hunk 13]
```diff
@@ -362,23 +373,62 @@ public void resetHoodEncoder(boolean reset) {
...
+            // angle = MIN_TURRET_ANGLE;
+            // }
+        }
+
+        if (loopToMax) {
+            angle = MAX_TURRET_ANGLE;
+            if (t.get() > 1 && isTurretAtAngle(MAX_TURRET_ANGLE)) {
+                loopToMax = false;
+                t.reset();
+                t.stop();
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystem/ShooterSubsystem.java:412]: Need to swap these, as t.stop() will simply set the timer to whatever time elapsed between these two lines of code, which may not be zero.  Ditto for below.

File: src/main/java/frc/team2412/robot/Hardware.java (hunk 1 above)
Comment (jamesdooley4) [reply src/main/java/frc/team2412/robot/Hardware.java:145]: Still crashing...

Comment (1198159) [reply src/main/java/frc/team2412/robot/Hardware.java:145]: fixed

File: src/main/java/frc/team2412/robot/commands/shooter/ShooterTargetCommand.java (hunk 4 above)
Comment (1198159) [reply src/main/java/frc/team2412/robot/commands/shooter/ShooterTargetCommand.java:44]: not using bitmap

File: src/main/java/frc/team2412/robot/subsystem/ShooterSubsystem.java (hunk 13 above)
Comment (1198159) [reply src/main/java/frc/team2412/robot/subsystem/ShooterSubsystem.java:412]: intended feature see lines above

**Comment by 1198159** [comment]
updated alot

**Comment by obj-obj** [comment]
Probably a good idea

**Pull request merged**